Add logical_not operator.

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24379 Implement operator logical_and
* #23978 Let logical_xor support non-bool tensors.
* #23916 Let logical_not support non-bool tensors.
* #23860 Recommend logical_not() instead of bitwise_not() when applying sub and neg on bool tensors.
* #23847 Add logical_xor operator
* **#23839 Add logical_not operator.**

Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

Differential Revision: [D16678301](https://our.internmc.facebook.com/intern/diff/D16678301)

xuhdev committed:
Add logical_not operator.

Close #23836

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

xuhdev committed:
Update on "Add logical_not operator."


Related to #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

gchanan commented on aten/src/ATen/Dispatch.h:
```diff
@@ -220,6 +220,23 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
     }                                                                        \
   }()
 
+#define AT_DISPATCH_INTEGRAL_TYPES_AND_BOOL(TYPE, NAME, ...)                 \
```
we've shied away from these crazily-named dispatch macros -- the ones that continue to be in here are here for BC reasons.

Can we call make a `AT_DISPATCH_INTEGRAL_TYPES_AND` macro and make it follow `AT_DISPATCH_ALL_TYPES_AND` (i.e. don't call `::detail::scalar_type`)?

xuhdev replied:
I just realized that this is actually not used in this PR. I have removed this macro.

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

gchanan commented on torch/_torch_docs.py:
```diff
@@ -2536,6 +2536,32 @@ def merge_dicts(*dicts):
+           r"""
+logical_not(input, out=None) -> Tensor
+
+Computes the element-wise logical NOT of the given input tensor. The output tensor always has the same dtype as that of
```
this is different than numpy's logical_not, which defaults to returning bool -- is there any reason why we don't do that?  Basically, I think this should behave the same way as our comparison operators.

xuhdev replied:
If we do this, can we still have inplace operator? What I'm thinking is that, in this case, a user should use `bool()` to do the conversion if a user needs bool tensors. This is also consistent with what `bitwise_not()` currently does.

gchanan replied:
I think it should be behave like a comparison ops; in that case inplace keeps the dtype:
```
>>> x.gt_(1)
tensor([[0., 0., 1.],
        [0., 0., 0.]])
```

btw if this is difficult it seems fine to just support boolean for now (so we can get it into 1.2) and then add the other dtypes in a follow-on patch.

xuhdev replied:
OK, I'll change it to bool only for now so we can safely have it into v1.2.

gchanan commented on torch/_torch_docs.py:
```diff
@@ -2536,6 +2536,32 @@ def merge_dicts(*dicts):
+Computes the element-wise logical NOT of the given input tensor. The output tensor always has the same dtype as that of
+:attr:`input`.
+
+- The logical NOT of a bool is the one that is the same as standard textbook definition: False -> True, True ->
```
I don't really think you need this section, the examples are sufficient.

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

gchanan commented on test/test_torch.py:
```diff
@@ -1819,6 +1819,25 @@ def _test_bitwise_not(self, device):
 
+    @staticmethod
+    def _test_logical_not(self, device):
+        for dtype in (torch.bool, torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64,
```
can you use https://github.com/pytorch/pytorch/blob/9588cd921efd051a982188220b8215914bd5790e/torch/testing/__init__.py#L94-L102?

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

gchanan commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -56,6 +56,27 @@ Tensor& bitwise_not_out(Tensor& result, const Tensor& self) {
+
+Tensor& logical_not_out(Tensor& result, const Tensor& self) {
+  TORCH_CHECK(self.scalar_type() == kBool,
+              "logical_not only supports bool tensors.");
```
nit: I'd say "only currently supports" to make it clear we are planning to add support for other types.

gchanan commented on test/test_torch.py:
```diff
@@ -1819,6 +1819,24 @@ def _test_bitwise_not(self, device):
+            self.assertEqual(expected_res, b)
+            # in-place
+            a.logical_not_()
+            self.assertEqual(expected_res, a)
```
what happens if you pass the wrong dtype as an `out` parameter?

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

gchanan commented:
clang tidy failure looks bogus.

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

Differential Revision: [D16678301](https://our.internmc.facebook.com/intern/diff/D16678301)

gchanan commented:
@xuhdev: are all your recent changes to this stack just merges/rebases?  I need to know if landing a prior version is safe.

also, can you stop merging/rebasing?  I already checked that the failures weren't related to this diff, and we have some internal systems that will reject commits if the version we are trying to merge isn't up-to-date with the the PR.  I'm not sure if the merges will go through now or I'll have to do redo it.

I realize we have no automated communication on this state so this is totally invisible to you, sorry about that!

xuhdev commented:
@gchanan Yes, I just rebased this morning, in case there's any conflict. Sorry for the confusion!

gchanan commented:
ok, I think it makes the most sense to punt on getting this into v1.2 -- mainly because we can't land anything right now.

gchanan commented:
Also, no need to apologize -- thanks so much for your contributions, they are really great.

xuhdev commented:
@gchanan No problem! Thanks for your review!

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

Differential Revision: [D16678301](https://our.internmc.facebook.com/intern/diff/D16678301)

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

Differential Revision: [D16678301](https://our.internmc.facebook.com/intern/diff/D16678301)

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

Differential Revision: [D16678301](https://our.internmc.facebook.com/intern/diff/D16678301)

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

Differential Revision: [D16678301](https://our.internmc.facebook.com/intern/diff/D16678301)

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

Differential Revision: [D16678301](https://our.internmc.facebook.com/intern/diff/D16678301)

xuhdev committed:
Update on "Add logical_not operator."


Close #23836

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23839

Differential Revision: [D16678301](https://our.internmc.facebook.com/intern/diff/D16678301)
